agent/testflinger_agent/job.py

```diff
@@ -27,239 +35,712 @@
+        if not self.cmd:
+            logger.info("No %s_command configured, skipping...", self.phase_id)
+            return False
+        return True
+
+    def run_core(self) -> JobPhaseResult:
+        # execute the external command
+        logger.info("Running %s_command: %s", self.phase_id, self.cmd)
+        try:
+            exit_code, event, detail = self.runner.run(self.cmd)
+            if exit_code == 0:
+                # complete, successful run
+                return JobPhaseResult(
+                    exit_code=exit_code,
+                    event=f"{self.phase_id}_success",
+                )
+            else:
+                # [NOTE] This is a deviation from the current approach
+                # where a separate event is emitted when stop events are
+                # returned from the runner and then a fail event on top
+                # Here we *only* emit the stop event
+                # self.emitter.emit_event(event, detail)
+                if event:
+                    return JobPhaseResult(
+                        exit_code=exit_code,
+                        event=event,
+                        detail=detail,
+                    )
+                else:
+                    return JobPhaseResult(
+                        exit_code=exit_code,
+                        event=f"{self.phase_id}_fail",
+                        detail=self.parse_error_logs(),
+                    )
+        except Exception as error:
+            # failed phase run due to an exception
+            detail = f"{type(error).__name__}: {error}"
+            logger.exception(detail)
+            return JobPhaseResult(
+                exit_code=100,
+                event=f"{self.phase_id}_fail",
+                detail=detail,
+            )
+
+    def parse_error_logs(self):
+        # [TODO] Move filenames used to pass information to the common module
+        with open(
+            self.params.rundir / "device-connector-error.json", "r"
+        ) as error_file:
+            error_file_contents = error_file.read()
+            try:
+                exception_info = json.loads(error_file_contents)[
+                    f"{self.phase_id}_exception_info"
+                ]
+                if exception_info["exception_cause"] is None:
+                    detail = "%s: %s" % (
+                        exception_info["exception_name"],
+                        exception_info["exception_message"],
+                    )
+                else:
+                    detail = "%s: %s caused by %s" % (
+                        exception_info["exception_name"],
+                        exception_info["exception_message"],
+                        exception_info["exception_cause"],
+                    )
+                return detail
+            except (json.JSONDecodeError, KeyError):
+                # [TODO] When do these errors arise?
+                return ""
+
+
+'''
```

Comment:
I'm sure you're coming back to clean this... just marking it. I know it's still draft :)